api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirEncounterDaoImpl.java

```diff
@@ -44,4 +57,121 @@ public Encounter getEncounterByUuid(String uuid) {
+		handleLocationReference(criteria, location);
+		handleParticipantReference(criteria, participant);
+		handleSubjectReference(criteria, subject);
+		//TODO: handle ReferenceParam for participant, subject, location
+		
+		return criteria.list();
+	}
+	
+	private void handleLocationReference(Criteria criteria, ReferenceParam locationReference) {
+		if (locationReference != null) {
+			criteria.createAlias("location", "l");
+			
+			if (locationReference.getChain() != null) {
+				switch (locationReference.getChain()) {
+					case Location.SP_ADDRESS_CITY:
+						criteria.add(ilike("l.cityVillage", locationReference.getValue(), MatchMode.START));
+						break;
+					case Location.SP_ADDRESS_STATE:
+						criteria.add(ilike("l.stateProvince", locationReference.getValue(), MatchMode.START));
+						break;
+					case Location.SP_ADDRESS_POSTALCODE:
+						criteria.add(ilike("l.postalCode", locationReference.getValue(), MatchMode.START));
+						break;
+					case Location.SP_ADDRESS_COUNTRY:
+						criteria.add(ilike("l.country", locationReference.getValue(), MatchMode.START));
+						break;
+					case "":
+						criteria.add(eq("l.uuid", locationReference.getValue()));
+						break;
+				}
+			}
+		}
+	}
+	
+	private void handleParticipantReference(Criteria criteria, ReferenceParam participantReference) {
+		if (participantReference != null) {
+			criteria.createAlias("encounterProviders", "ep");
+			
+			if (participantReference.getChain() != null) {
+				switch (participantReference.getChain()) {
+					case Patient.SP_IDENTIFIER:
+						criteria.createAlias("ep.provider", "p").add(ilike("p.identifier", participantReference.getValue()));
+						break;
+					case Patient.SP_GIVEN:
+						criteria.createAlias("ep.provider", "p").createAlias("p.person", "ps").createAlias("ps.names", "pn")
+						        .add(ilike("pn.givenName", participantReference.getValue(), MatchMode.START));
+						break;
+					case Patient.SP_FAMILY:
+						criteria.createAlias("ep.provider", "p").createAlias("p.person", "ps").createAlias("ps.names", "pn")
+						        .add(ilike("pn.familyName", participantReference.getValue(), MatchMode.START));
+						break;
+					case Patient.SP_NAME:
+						criteria.createAlias("ep.provider", "p").createAlias("p.person", "ps").createAlias("ps.names", "pn");
+						
+						List<Optional<Criterion>> criterionList = new ArrayList<>();
+						
+						for (String token : StringUtils.split(participantReference.getValue(), " \t,")) {
+							criterionList.add(propertyLike("pn.givenName", token));
+							criterionList.add(propertyLike("pn.middleName", token));
+							criterionList.add(propertyLike("pn.familyName", token));
+						}
+						
+						criteria.add(or(toCriteriaArray(criterionList)));
+						break;
+					case "":
+						criteria.add(eq("ep.uuid", participantReference.getValue()));
+						break;
+				}
+			}
+		}
+	}
+	
+	protected void handleSubjectReference(Criteria criteria, ReferenceParam subjectReference) {
```

Comment:
How does this differ from the `handlePatientReference()` function that already exists?

Reply:
@ibacher `handlePatientReference()` has `person` as alias but this has `patient` as alias if I try to use `handlePatientReference()` then an erreo like cannot resolve propery person pops (after writing tests)

Reply:
Absolutely right. I've fixed this in the latest master.

Reply:
@VaishSiddharth Can you rebase onto the latest version of master and use the `handlePatientReference()` method instead? Thanks 😄

Reply:
@ibacher Working in it. Will soon send another commit after resolving all these issues.